src/QMCWaveFunctions/Determinant.h

```diff
@@ -0,0 +1,245 @@
+    }
+  }
+#pragma omp master
+  std::cout << tag << " Identity Error = " << error / nrows / nrows
+            << std::endl;
+}
+
+// FIXME do we want to keep this in the miniapp?
+template <typename MT1, typename MT2>
+void checkDiff(const MT1 &a, const MT2 &b, const std::string &tag)
+{
+  const int nrows = a.rows();
+  const int ncols = a.cols();
+  constexpr double czero(0.0);
+  double error = czero;
+  for (int i = 0; i < nrows; ++i)
+    for (int j = 0; j < ncols; ++j)
+      error += std::abs(static_cast<double>(a(i, j) - b(i, j)));
+#pragma omp master
+  std::cout << tag << " diff Error = " << error / nrows / nrows << std::endl;
+}
+
+struct DiracDeterminant : public DiracDeterminantBase
+{
+  DiracDeterminant(int nels, RandomGenerator<double> RNG)
+  {
+    psiMinv.resize(nels, nels);
+    psiV.resize(nels);
+    psiM.resize(nels, nels);
+
+    pivot.resize(nels);
+    psiMsave.resize(nels, nels);
+
+    // now we "void initialize(RandomGenerator<T> RNG)"
+
+    nels = psiM.rows();
+    // get lwork and resize workspace
+    LWork = getGetriWorkspace(psiM.data(), nels, nels, pivot.data());
+    work.resize(LWork);
+
+    myRandom = RNG;
+    constexpr double shift(0.5);
+    RNG.generate_uniform(psiMsave.data(), nels * nels);
+    psiMsave -= shift;
+
+    double phase;
+    transpose(psiMsave.data(), psiM.data(), nels, nels);
+    LogValue = InvertWithLog(psiM.data(), nels, nels, work.data(), LWork,
+                             pivot.data(), phase);
+    std::copy_n(psiM.data(), nels * nels, psiMinv.data());
+
+    if (omp_get_num_threads() == 1)
+    {
+      checkIdentity(psiMsave, psiM, "Psi_0 * psiM(double)");
+      checkIdentity(psiMsave, psiMinv, "Psi_0 * psiMinv(T)");
+      checkDiff(psiM, psiMinv, "psiM(double)-psiMinv(T)");
+    }
+  }
+
+  /// recompute the inverse
+  inline void recompute()
+  {
+    const int nels = psiV.size();
+    transpose(psiMsave.data(), psiM.data(), nels, nels);
+    InvertOnly(psiM.data(), nels, nels, work.data(), pivot.data(), LWork);
+    std::copy_n(psiM.data(), nels * nels, psiMinv.data());
+  }
+
+  /** return determinant ratio for the row replacement
+   * @param iel the row (active particle) index
+   */
+  inline double ratio(int iel)
+  {
+    const int nels = psiV.size();
+    constexpr double shift(0.5);
+    constexpr double czero(0);
+    for (int j = 0; j < nels; ++j)
+      psiV[j] = myRandom() - shift;
+    curRatio = inner_product_n(psiV.data(), psiMinv[iel], nels, czero);
+    return curRatio;
+  }
+
+  /** accept the row and update the inverse */
+  inline void acceptMove(int iel)
+  {
+    const int nels = psiV.size();
+    updateRow(psiMinv.data(), psiV.data(), nels, nels, iel, curRatio);
+    std::copy_n(psiV.data(), nels, psiMsave[iel]);
+  }
+
+  /** accessor functions for checking */
+  inline double operator()(int i) const { return psiMinv(i); }
+  inline int size() const { return psiMinv.size(); }
+
+private:
+  /// log|det|
+  double LogValue;
```

Comment:
I prefer the whole class not having hard coded doubles. We should support mixed precision just like the others.